emsdk_env.sh

```diff
@@ -1,28 +1,64 @@
 # This script is sourced by the user and uses
-# their shell. Try not to use bashisms.
-
+# their shell.
+#
+# This script tries to find its location but
+# this does not work in every shell.
+#
+# It is known to work in bash, zsh and ksh
+#
 # Do not execute this script without sourcing,
 # because it won't have any effect then.
 # That is, always run this script with
 #
-#     . ./emsdk_env.sh
+#     . /path/to/emsdk_env.sh
+#
 # or
-#     source ./emsdk_env.sh
+#
+#     source /path/to/emsdk_env.sh
 #
 # instead of just plainly running with
 #
 #     ./emsdk_env.sh
 #
 # which won't have any effect.
-if [ -z "$BASH_SOURCE" ]; then
-  if [ ! -f "./emsdk.py" ]; then
-    echo "error: You must be in the same directory as emsdk_env.sh when sourcing it (or switch to the bash shell)" 1>&2
+
+CURRENT_SCRIPT=
+DIR="."
+
+if [ -n "$BASH_SOURCE" ]; then
+  CURRENT_SCRIPT="$BASH_SOURCE"
+elif [ -n "$ZSH_VERSION" ]; then
+  CURRENT_SCRIPT="${(%):-%x}"
+elif [ -n "$KSH_VERSION" ]; then
+  CURRENT_SCRIPT=${.sh.file}
+fi
+
+if [ -n "$CURRENT_SCRIPT" ]; then
+  DIR=$(dirname "$CURRENT_SCRIPT")
+  # now work out actual DIR if this is part of a symlink
+  if [ -h "$CURRENT_SCRIPT" ]; then
+    SYMDIR=$(dirname "$(ls -l "$CURRENT_SCRIPT" | sed -n "s/.*-> //p")")
+    if [ -z "$SYMDIR" ]; then
+      SYMDIR="."
+    fi
+    FULLDIR="$DIR/$SYMDIR"
+    DIR=$(cd "$FULLDIR" > /dev/null 2>&1; /bin/pwd)
+    unset SYMDIR
+    unset FULLDIR
   fi
-  DIR="."
-else
-  DIR="$(dirname "$BASH_SOURCE")"
+fi
+
+if [ ! -f "$DIR/emsdk.py" ]; then
+  echo "Error: unable to determine 'emsdk' directory. Perhaps you are using a shell or" 1>&2
+  echo "       environment that this script does not support." 1>&2
+  echo 1>&2
+  echo "A possible solution is to source this script while in the 'emsdk' directory." 1>&2
+  echo 1>&2
+  unset DIR
+  return
```

Comment:
What does return do in this position?

Reply:
this basically returns early if the 'eval' below is going to fail.

prevents the error that you get on the last line:

```
$ . ../emsdk_env.sh
Error: unable to determine 'emsdk' directory. Perhaps you are using a shell or
       environment that this script does not support.

A possible solution is to source this script while in the 'emsdk' directory.

dash: 1: ../emsdk_env.sh: /emsdk: not found
```

so instead you just get:

```
$ . ../emsdk_env.sh
Error: unable to determine 'emsdk' directory. Perhaps you are using a shell or
       environment that this script does not support.

A possible solution is to source this script while in the 'emsdk' directory.
```

not that big of a deal, either way. Just seemed cleaner, but then I'm not shell programming expert at all.